Several Integer overflow problems in syscall `mmap`/`munmap`/`mprotect`/`madvise` caused by large `size`

### Describe the bug
There is a add with overflow error in `find_free_region()` at kernel/aster-nix/src/vm/vmar/mod.rs:167 when make a `mmap` syscall with large `size`.

https://github.com/asterinas/asterinas/blob/562e64437584279783f244edba10b512beddc81d/kernel/aster-nix/src/vm/vmar/mod.rs#L167


### To Reproduce
1. Compile a program which calls `mmap`:
```C
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <unistd.h>

int main() {
  void *addr = mmap(NULL, 0xfffffffffffff000, 0x3, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);

  return 0;
}

```
2. Run the compiled program in Asterinas.

### Expected behavior
Asterinas reports `attempt to add with overflow` panic and is terminated.

### Environment
- Official docker asterinas/asterinas:0.6.2
- 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
- Asterinas version: main 562e6443

### Logs
```
~ # /root/munmap.c 
panicked at /root/asterinas/kernel/aster-nix/src/vm/vmar/mod.rs:167:43:
attempt to add with overflow
Printing stack trace:
   1: fn 0xffffffff888695b0 - pc 0xffffffff888695c8 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a7a70;

   2: fn 0xffffffff88869390 - pc 0xffffffff88869508 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a7a80;

   3: fn 0xffffffff88049000 - pc 0xffffffff8804900a / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a7c00;

   4: fn 0xffffffff889a8690 - pc 0xffffffff889a8712 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a7c10;

   5: fn 0xffffffff88983e50 - pc 0xffffffff88983e7e / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a7ca0;

   6: fn 0xffffffff88470450 - pc 0xffffffff884707f7 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a7cf0;

   7: fn 0xffffffff884742b0 - pc 0xffffffff88474c1a / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a7f30;

   8: fn 0xffffffff8855aa00 - pc 0xffffffff8855ab86 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a8900;

   9: fn 0xffffffff8855b6d0 - pc 0xffffffff8855b824 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a8dd0;

  10: fn 0xffffffff8844e2f0 - pc 0xffffffff8844f4da / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a90f0;

  11: fn 0xffffffff8844e000 - pc 0xffffffff8844e168 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a9da0;

  12: fn 0xffffffff8856ded0 - pc 0xffffffff88572b3c / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a9f30;

  13: fn 0xffffffff880df4c0 - pc 0xffffffff880df54e / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4c03d0;

  14: fn 0xffffffff882db7a0 - pc 0xffffffff882dc31f / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4c0570;

  15: fn 0xffffffff883ffc40 - pc 0xffffffff883ffc4e / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4c0f90;

  16: fn 0xffffffff887c0770 - pc 0xffffffff887c0786 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4c0fb0;

  17: fn 0xffffffff887a61e0 - pc 0xffffffff887a6249 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4c0fd0;

  18: fn                0x0 - pc                0x0 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4c1000;

[OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) /root/asterinas/ostd/src/panicking.rs:107
(  2) /root/asterinas/ostd/src/panicking.rs:59
(  3) 0yfjuqijmdltwb3g9ceg2tyxs:?
(  4) ??:?
(  5) ??:?
(  6) axbvs75vp1k66nwsaobv39t4g:?
(  7) /root/asterinas/kernel/aster-nix/src/vm/vmar/mod.rs:607
(  8) /root/asterinas/kernel/aster-nix/src/vm/vmar/vm_mapping.rs:102
(  9) /root/asterinas/kernel/aster-nix/src/vm/vmar/vm_mapping.rs:695
( 10) /root/asterinas/kernel/aster-nix/src/syscall/mmap.rs:137
( 11) /root/asterinas/kernel/aster-nix/src/syscall/mmap.rs:33
( 12) /root/asterinas/kernel/aster-nix/src/syscall/mod.rs:220
( 13) /root/asterinas/kernel/aster-nix/src/syscall/mod.rs:328
( 14) /root/asterinas/kernel/aster-nix/src/thread/task.rs:69
( 15) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:79
( 16) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2077
( 17) /root/asterinas/ostd/src/task/task/mod.rs:310
```

Similar bugs are also found in `munmap`/`mprotect`/`madvise`:

https://github.com/asterinas/asterinas/blob/562e64437584279783f244edba10b512beddc81d/kernel/aster-nix/src/syscall/munmap.rs#L13

<details>

<summary> log info of munmap </summary>

```
~ # /root/munmap.c 
panicked at /root/asterinas/kernel/aster-nix/src/syscall/munmap.rs:16:29:
attempt to add with overflow
Printing stack trace:
   1: fn 0xffffffff888695b0 - pc 0xffffffff888695c8 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a99c0;

   2: fn 0xffffffff88869390 - pc 0xffffffff88869508 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a99d0;

   3: fn 0xffffffff88049000 - pc 0xffffffff8804900a / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a9b50;

   4: fn 0xffffffff889a8690 - pc 0xffffffff889a8712 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a9b60;

   5: fn 0xffffffff88983e50 - pc 0xffffffff88983e7e / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a9bf0;

   6: fn 0xffffffff883b6b10 - pc 0xffffffff883b6ffc / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a9c40;

   7: fn 0xffffffff8856ded0 - pc 0xffffffff885734f2 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4a9f30;

   8: fn 0xffffffff880df4c0 - pc 0xffffffff880df54e / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4c03d0;

   9: fn 0xffffffff882db7a0 - pc 0xffffffff882dc31f / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4c0570;

  10: fn 0xffffffff883ffc40 - pc 0xffffffff883ffc4e / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4c0f90;

  11: fn 0xffffffff887c0770 - pc 0xffffffff887c0786 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4c0fb0;

  12: fn 0xffffffff887a61e0 - pc 0xffffffff887a6249 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4c0fd0;

  13: fn                0x0 - pc                0x0 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e4c1000;

[OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) /root/asterinas/ostd/src/panicking.rs:107
(  2) /root/asterinas/ostd/src/panicking.rs:59
(  3) 0yfjuqijmdltwb3g9ceg2tyxs:?
(  4) ??:?
(  5) ??:?
(  6) 8j5o2srz7f6ei3ofqjuqjpq09:?
(  7) /root/asterinas/kernel/aster-nix/src/syscall/mod.rs:164
(  8) /root/asterinas/kernel/aster-nix/src/syscall/mod.rs:328
(  9) /root/asterinas/kernel/aster-nix/src/thread/task.rs:69
( 10) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:79
( 11) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2077
( 12) /root/asterinas/ostd/src/task/task/mod.rs:310
make: *** [Makefile:153: run] Error 1
```
</details>


https://github.com/asterinas/asterinas/blob/562e64437584279783f244edba10b512beddc81d/kernel/aster-nix/src/syscall/mprotect.rs#L17


<details>

<summary> log info of mprotect </summary>

```
~ # /root/mprotect.c 
panicked at /root/asterinas/kernel/aster-nix/src/syscall/mprotect.rs:23:23:
attempt to add with overflow
Printing stack trace:
   1: fn 0xffffffff888695b0 - pc 0xffffffff888695c8 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5a9a40;

   2: fn 0xffffffff88869390 - pc 0xffffffff88869508 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5a9a50;

   3: fn 0xffffffff88049000 - pc 0xffffffff8804900a / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5a9bd0;

   4: fn 0xffffffff889a8690 - pc 0xffffffff889a8712 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5a9be0;

   5: fn 0xffffffff88983e50 - pc 0xffffffff88983e7e / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5a9c70;

   6: fn 0xffffffff8855a540 - pc 0xffffffff8855a91d / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5a9cc0;

   7: fn 0xffffffff8856ded0 - pc 0xffffffff88573019 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5a9f30;

   8: fn 0xffffffff880df4c0 - pc 0xffffffff880df54e / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5c03d0;

   9: fn 0xffffffff882db7a0 - pc 0xffffffff882dc31f / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5c0570;

  10: fn 0xffffffff883ffc40 - pc 0xffffffff883ffc4e / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5c0f90;

  11: fn 0xffffffff887c0770 - pc 0xffffffff887c0786 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5c0fb0;

  12: fn 0xffffffff887a61e0 - pc 0xffffffff887a6249 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5c0fd0;

  13: fn                0x0 - pc                0x0 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e5c1000;

[OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) /root/asterinas/ostd/src/panicking.rs:107
(  2) /root/asterinas/ostd/src/panicking.rs:59
(  3) 0yfjuqijmdltwb3g9ceg2tyxs:?
(  4) ??:?
(  5) ??:?
(  6) d9wc6hqkevrkh55mc6uw1tmac:?
(  7) /root/asterinas/kernel/aster-nix/src/syscall/mod.rs:171
(  8) /root/asterinas/kernel/aster-nix/src/syscall/mod.rs:328
(  9) /root/asterinas/kernel/aster-nix/src/thread/task.rs:69
( 10) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:79
( 11) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2077
( 12) /root/asterinas/ostd/src/task/task/mod.rs:310
make: *** [Makefile:153: run] Error 1
```
</details>


https://github.com/asterinas/asterinas/blob/562e64437584279783f244edba10b512beddc81d/kernel/aster-nix/src/syscall/madvise.rs#L48

<details>

<summary> log info of madvise </summary>

```
~ # /root/madvise.c 
panicked at /root/asterinas/kernel/aster-nix/src/syscall/madvise.rs:48:32:
attempt to add with overflow
Printing stack trace:
   1: fn 0xffffffff888695b0 - pc 0xffffffff888695c8 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7a9880;

   2: fn 0xffffffff88869390 - pc 0xffffffff88869508 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7a9890;

   3: fn 0xffffffff88049000 - pc 0xffffffff8804900a / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7a9a10;

   4: fn 0xffffffff889a8690 - pc 0xffffffff889a8712 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7a9a20;

   5: fn 0xffffffff88983e50 - pc 0xffffffff88983e7e / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7a9ab0;

   6: fn 0xffffffff88534cd0 - pc 0xffffffff88534d63 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7a9b00;

   7: fn 0xffffffff885344a0 - pc 0xffffffff88534b58 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7a9b80;

   8: fn 0xffffffff8856ded0 - pc 0xffffffff88577deb / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7a9f30;

   9: fn 0xffffffff880df4c0 - pc 0xffffffff880df54e / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7c03d0;

  10: fn 0xffffffff882db7a0 - pc 0xffffffff882dc31f / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7c0570;

  11: fn 0xffffffff883ffc40 - pc 0xffffffff883ffc4e / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7c0f90;

  12: fn 0xffffffff887c0770 - pc 0xffffffff887c0786 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7c0fb0;

  13: fn 0xffffffff887a61e0 - pc 0xffffffff887a6249 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7c0fd0;

  14: fn                0x0 - pc                0x0 / registers:

     rax               0x12; rdx 0xffffffff88a38470; rcx                0x1; rbx                0x0;
     rsi                0x0; rdi                0x0; rbp                0x0; rsp 0xffff80007e7c1000;

[OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) /root/asterinas/ostd/src/panicking.rs:107
(  2) /root/asterinas/ostd/src/panicking.rs:59
(  3) 0yfjuqijmdltwb3g9ceg2tyxs:?
(  4) ??:?
(  5) ??:?
(  6) cgf2wirtt588u1sy2ej5ns7nw:?
(  7) /root/asterinas/kernel/aster-nix/src/syscall/madvise.rs:40
(  8) /root/asterinas/kernel/aster-nix/src/syscall/mod.rs:171
(  9) /root/asterinas/kernel/aster-nix/src/syscall/mod.rs:328
( 10) /root/asterinas/kernel/aster-nix/src/thread/task.rs:69
( 11) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:79
( 12) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2077
( 13) /root/asterinas/ostd/src/task/task/mod.rs:310
make: *** [Makefile:153: run] Error 1
```
</details>

--- a/kernel/src/syscall/madvise.rs
+++ b/kernel/src/syscall/madvise.rs
@@ -20,7 +20,7 @@ pub fn sys_madvise(
     if start % PAGE_SIZE != 0 {
         return_errno_with_message!(Errno::EINVAL, "the start address should be page aligned");
     }
-    if len == 0 {
+    if len == 0 || len > usize::MAX - PAGE_SIZE + 1 {
         return Ok(SyscallReturn::Return(0));
     }
 

--- a/kernel/src/syscall/madvise.rs
+++ b/kernel/src/syscall/madvise.rs
@@ -20,7 +20,10 @@ pub fn sys_madvise(
     if start % PAGE_SIZE != 0 {
         return_errno_with_message!(Errno::EINVAL, "the start address should be page aligned");
     }
-    if len == 0 || len > usize::MAX - PAGE_SIZE + 1 {
+    if len > usize::MAX - PAGE_SIZE + 1 {
+        return_errno_with_message!(Errno::EINVAL, "len align overflow");
+    }
+    if len == 0 {
         return Ok(SyscallReturn::Return(0));
     }
 

--- a/kernel/src/syscall/madvise.rs
+++ b/kernel/src/syscall/madvise.rs
@@ -20,7 +20,7 @@ pub fn sys_madvise(
     if start % PAGE_SIZE != 0 {
         return_errno_with_message!(Errno::EINVAL, "the start address should be page aligned");
     }
-    if len > usize::MAX - PAGE_SIZE + 1 {
+    if len > isize::MAX as usize {
         return_errno_with_message!(Errno::EINVAL, "len align overflow");
     }
     if len == 0 {

--- a/kernel/src/syscall/madvise.rs
+++ b/kernel/src/syscall/madvise.rs
@@ -25,6 +25,10 @@ pub fn sys_madvise(
     }
 
     let len = len.align_up(PAGE_SIZE);
+    let end = start.checked_add(len).ok_or(Error::with_message(
+        Errno::EINVAL,
+        "integer overflow when (start + len)",
+    ))?;
     match behavior {
         MadviseBehavior::MADV_NORMAL
         | MadviseBehavior::MADV_SEQUENTIAL
@@ -37,15 +41,15 @@ pub fn sys_madvise(
         MadviseBehavior::MADV_DONTNEED => {
             warn!("MADV_DONTNEED isn't implemented, do nothing for now.");
         }
-        MadviseBehavior::MADV_FREE => madv_free(start, len, ctx)?,
+        MadviseBehavior::MADV_FREE => madv_free(start, end, ctx)?,
         _ => todo!(),
     }
     Ok(SyscallReturn::Return(0))
 }
 
-fn madv_free(start: Vaddr, len: usize, ctx: &Context) -> Result<()> {
+fn madv_free(start: Vaddr, end: Vaddr, ctx: &Context) -> Result<()> {
     let root_vmar = ctx.process.root_vmar();
-    let advised_range = start..start + len;
+    let advised_range = start..end;
     let _ = root_vmar.destroy(advised_range);
 
     Ok(())

--- a/kernel/src/syscall/mmap.rs
+++ b/kernel/src/syscall/mmap.rs
@@ -57,6 +57,9 @@ fn do_sys_mmap(
     if len == 0 {
         return_errno_with_message!(Errno::EINVAL, "mmap len cannot be zero");
     }
+    if len > usize::MAX - PAGE_SIZE + 1 {
+        return_errno_with_message!(Errno::ENOMEM, "mmap len align overflow");
+    }
 
     let len = len.align_up(PAGE_SIZE);
 

--- a/kernel/src/syscall/mmap.rs
+++ b/kernel/src/syscall/mmap.rs
@@ -57,14 +57,21 @@ fn do_sys_mmap(
     if len == 0 {
         return_errno_with_message!(Errno::EINVAL, "mmap len cannot be zero");
     }
-    if len > usize::MAX - PAGE_SIZE + 1 {
-        return_errno_with_message!(Errno::ENOMEM, "mmap len align overflow");
+    if len > isize::MAX as usize {
+        return_errno_with_message!(Errno::ENOMEM, "mmap len too large");
     }
 
     let len = len.align_up(PAGE_SIZE);
 
     if offset % PAGE_SIZE != 0 {
         return_errno_with_message!(Errno::EINVAL, "mmap only support page-aligned offset");
+    }
+    offset.checked_add(len).ok_or(Error::with_message(
+        Errno::EOVERFLOW,
+        "integer overflow when (offset + len)",
+    ))?;
+    if addr > isize::MAX as usize - len {
+        return_errno_with_message!(Errno::ENOMEM, "mmap (addr + len) too large");
     }
 
     let root_vmar = ctx.process.root_vmar();

--- a/kernel/src/syscall/mprotect.rs
+++ b/kernel/src/syscall/mprotect.rs
@@ -22,6 +22,9 @@ pub fn sys_mprotect(addr: Vaddr, len: usize, perms: u64, ctx: &Context) -> Resul
     if len == 0 {
         return Ok(SyscallReturn::Return(0));
     }
+    if len > usize::MAX - PAGE_SIZE + 1 {
+        return_errno_with_message!(Errno::ENOMEM, "len align overflow");
+    }
 
     let len = len.align_up(PAGE_SIZE);
     let end = addr.checked_add(len).ok_or(Error::with_message(

--- a/kernel/src/syscall/mprotect.rs
+++ b/kernel/src/syscall/mprotect.rs
@@ -24,7 +24,11 @@ pub fn sys_mprotect(addr: Vaddr, len: usize, perms: u64, ctx: &Context) -> Resul
     }
 
     let len = len.align_up(PAGE_SIZE);
-    let range = addr..(addr + len);
+    let end = addr.checked_add(len).ok_or(Error::with_message(
+        Errno::ENOMEM,
+        "integer overflow when (addr + len)",
+    ))?;
+    let range = addr..end;
     root_vmar.protect(vm_perms, range)?;
     Ok(SyscallReturn::Return(0))
 }

--- a/kernel/src/syscall/mprotect.rs
+++ b/kernel/src/syscall/mprotect.rs
@@ -22,7 +22,7 @@ pub fn sys_mprotect(addr: Vaddr, len: usize, perms: u64, ctx: &Context) -> Resul
     if len == 0 {
         return Ok(SyscallReturn::Return(0));
     }
-    if len > usize::MAX - PAGE_SIZE + 1 {
+    if len > isize::MAX as usize {
         return_errno_with_message!(Errno::ENOMEM, "len align overflow");
     }
 

--- a/kernel/src/syscall/munmap.rs
+++ b/kernel/src/syscall/munmap.rs
@@ -14,6 +14,9 @@ pub fn sys_munmap(addr: Vaddr, len: usize, ctx: &Context) -> Result<SyscallRetur
     if len == 0 {
         return_errno_with_message!(Errno::EINVAL, "munmap len cannot be zero");
     }
+    if len > usize::MAX - PAGE_SIZE + 1 {
+        return_errno_with_message!(Errno::ENOMEM, "munmap len align overflow");
+    }
 
     let root_vmar = ctx.process.root_vmar();
     let len = len.align_up(PAGE_SIZE);

--- a/kernel/src/syscall/munmap.rs
+++ b/kernel/src/syscall/munmap.rs
@@ -17,7 +17,11 @@ pub fn sys_munmap(addr: Vaddr, len: usize, ctx: &Context) -> Result<SyscallRetur
 
     let root_vmar = ctx.process.root_vmar();
     let len = len.align_up(PAGE_SIZE);
-    debug!("unmap range = 0x{:x} - 0x{:x}", addr, addr + len);
-    root_vmar.destroy(addr..addr + len)?;
+    let end = addr.checked_add(len).ok_or(Error::with_message(
+        Errno::EINVAL,
+        "integer overflow when (addr + len)",
+    ))?;
+    debug!("unmap range = 0x{:x} - 0x{:x}", addr, end);
+    root_vmar.destroy(addr..end)?;
     Ok(SyscallReturn::Return(0))
 }

--- a/kernel/src/syscall/munmap.rs
+++ b/kernel/src/syscall/munmap.rs
@@ -14,7 +14,7 @@ pub fn sys_munmap(addr: Vaddr, len: usize, ctx: &Context) -> Result<SyscallRetur
     if len == 0 {
         return_errno_with_message!(Errno::EINVAL, "munmap len cannot be zero");
     }
-    if len > usize::MAX - PAGE_SIZE + 1 {
+    if len > isize::MAX as usize {
         return_errno_with_message!(Errno::ENOMEM, "munmap len align overflow");
     }
 

--- a/kernel/src/vm/vmar/mod.rs
+++ b/kernel/src/vm/vmar/mod.rs
@@ -169,7 +169,7 @@ impl VmarInner {
                         .checked_add(child_size)
                         .ok_or(Error::with_message(
                             Errno::ENOMEM,
-                            "integer overflow whem (child_vmar_real_start + child_size)",
+                            "integer overflow when (child_vmar_real_start + child_size)",
                         ))?;
                 if region_start <= child_vmar_real_start && child_vmar_real_end <= region_end {
                     return Ok((*region_base, child_vmar_real_start));

--- a/kernel/src/vm/vmar/mod.rs
+++ b/kernel/src/vm/vmar/mod.rs
@@ -164,7 +164,13 @@ impl VmarInner {
                 let region_start = free_region.start();
                 let region_end = free_region.end();
                 let child_vmar_real_start = region_start.align_up(align);
-                let child_vmar_real_end = child_vmar_real_start + child_size;
+                let child_vmar_real_end =
+                    child_vmar_real_start
+                        .checked_add(child_size)
+                        .ok_or(Error::with_message(
+                            Errno::ENOMEM,
+                            "integer overflow whem (child_vmar_real_start + child_size)",
+                        ))?;
                 if region_start <= child_vmar_real_start && child_vmar_real_end <= region_end {
                     return Ok((*region_base, child_vmar_real_start));
                 }

